pkg/pattern/drain/line_tokenizer.go

```diff
@@ -131,3 +174,115 @@ func (splittingTokenizer) Join(tokens []string, state interface{}) string {
 	}
 	return strBuilder.String()
 }
+
+func (splittingTokenizer) Clone(tokens []string, state interface{}) ([]string, interface{}) {
+	res := make([]string, len(tokens))
+	for i, token := range tokens {
+		res[i] = strings.Clone(token)
+	}
+	if state == nil {
+		return res, nil
+	}
+	spacesAfter := state.([]int)
+	spacesAfterCopy := make([]int, len(spacesAfter))
+	copy(spacesAfterCopy, spacesAfter)
+	return res, spacesAfterCopy
+}
+
+type logfmtTokenizer struct {
+	dec        *logfmt.Decoder
+	varReplace string
+}
+
+func newLogfmtTokenizer(varReplace string) *logfmtTokenizer {
+	return &logfmtTokenizer{
+		dec:        logfmt.NewDecoder(nil),
+		varReplace: varReplace,
+	}
+}
+
+func (t *logfmtTokenizer) Tokenize(line string, tokens []string, _ interface{}) ([]string, interface{}) {
+	if cap(tokens) == 0 {
+		tokens = make([]string, 0, 64)
+	}
+	tokens = tokens[:0]
+	t.dec.Reset(unsafeBytes(line))
+	for !t.dec.EOL() && t.dec.ScanKeyval() {
+		key := t.dec.Key()
+		if isVariableField(key) {
+			tokens = append(tokens, unsafeString(t.dec.Key()), t.varReplace)
```

Comment:
Interesting that this works - I thought that ScanKeyval reused those values after every call but it seems tthat it doesn't?

Reply:
It returns either a copy if it needs to decode quoted bytes or just a subslice.